### ISSUE TITLE
local variable 'cbl' referenced before assignment

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2476]
|<img alt="jack's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| jack reported|
|-|-|
|Trac ID|trac#2476|
|Type|defect|
|Created|2007-02-24 03:08:05Z|

Running twisted.words inside an XMPP component using the patches from [#11397](https://github.com/twisted/twisted/issues/11397) this Traceback happens occasionally:

```
        Traceback (most recent call last):
          File "/Users/jack/Sources/white/lib/python/twisted/python/log.py", line 48, in callWithLogger
            return callWithContext({"system": lp}, func, *args, **kw)
          File "/Users/jack/Sources/white/lib/python/twisted/python/log.py", line 33, in callWithContext
            return context.call({ILogContext: newCtx}, func, *args, **kw)
          File "/Users/jack/Sources/white/lib/python/twisted/python/context.py", line 59, in callWithContext
            return self.currentContext().callWithContext(ctx, func, *args, **kw)
          File "/Users/jack/Sources/white/lib/python/twisted/python/context.py", line 37, in callWithContext
            return func(*args,**kw)
        --- <exception caught here> ---
          File "/Users/jack/Sources/white/lib/python/twisted/internet/selectreactor.py", line 139, in _doReadOrWrite
            why = getattr(selectable, method)()
          File "/Users/jack/Sources/white/lib/python/twisted/internet/tcp.py", line 362, in doRead
            return self.protocol.dataReceived(data)
          File "/Users/jack/Sources/white/lib/python/twisted/words/xish/xmlstream.py", line 72, in dataReceived
            self.stream.parse(data)
          File "/Users/jack/Sources/white/lib/python/twisted/words/xish/domish.py", line 730, in parse
            self.parser.Parse(buffer)
          File "/Users/jack/Sources/white/lib/python/twisted/words/xish/domish.py", line 761, in _onStartElement
            self.DocumentStartEvent(e)
          File "/Users/jack/Sources/white/lib/python/twisted/words/protocols/jabber/xmlstream.py", line 543, in onDocumentStart
            self.onStreamError)
          File "/Users/jack/Sources/white/lib/python/twisted/words/xish/utility.py", line 179, in addOnetimeObserver
            self._addObserver(True, event, observerfn, priority, *args, **kwargs)
          File "/Users/jack/Sources/white/lib/python/twisted/words/xish/utility.py", line 222, in _addObserver
            cbl.addCallback(onetime, observerfn, *args, **kwargs)
        exceptions.UnboundLocalError: local variable 'cbl' referenced before assignment
```

The attached patch is a work around, but may not be a proper fix.

Attachments:

* [utility.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/277/2773b3399512236d16b9be785202a7232ed84150/b76ea63eada9a62a18c1be01837a0ea65f5d595e.diff) (831 bytes) - added by jack on 2007-02-24 03:08:37Z - workaround patch for the bug

<details><summary>Searchable metadata</summary>

```
trac-id__2476 2476
type__defect defect
reporter__jack jack
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__words words
keywords__xish_regression xish regression
time__1172286485000000 1172286485000000
changetime__1173625958000000 1173625958000000
version__None None
owner__ 
cc__ralphm cc__twonds cc__jack
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to @ralphm|
|-|-|

[#11397](https://github.com/twisted/twisted/issues/11397) should probably be backed out, test coverage for the changed code improved, and this bug fixed there.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_2"></a>|@ralphm commented|
|-|-|

(In [19705]) Rebranch for reverted r19669 re [#11397](https://github.com/twisted/twisted/issues/11397) and [#2476](https://github.com/twisted/twisted/issues/2476).

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_3"></a>|@ralphm commented|
|-|-|

The previous rebranch misfired, but this problem is now solved in branches:xish-dispatcher-1652-2. Moving over there for review.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_4"></a>|@ralphm set status to `closed`|
|-|-|

(In [19807]) Cleanup empty callback lists in EventDispatcher.

Author: ralphm
Reviewer: therve, exarkun
Fixes [#11397](https://github.com/twisted/twisted/issues/11397), [[#2476](https://github.com/twisted/twisted/issues/2476)](https://github.com/twisted/twisted/issues/2476).

EventDispatcher did not remove empty callback lists in the event callback
registration. This causes events to be matched to queries that will not
trigger any callback being called, causing significant slowdown when
many one-time callbacks are registered.

This change makes sure empty callback lists are cleaned out, and also optimizes
the internal bookkeeping to not have to use sorting on every callback
registration. Instead the observers are kept in different dictionaries (query
-> callback list) for each priority level. Then, upon event dispatch, the
observers priorities are sorted and then each associated dictionary is
processed in that order. This assumes the number of used priority levels is
very low to be efficient.

This remerge also fixes a regression introduced in r19669 that made multiple
observers for the same query fail (#2476).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
